### PR TITLE
Persistence Query: InmemReadJournal

### PR DESCRIPTION
**Pull request by cchepelov**
InmemJournal, whose purpose is testing, lacked a facility to read events like other journals (Leveldb, SQL/JDBC, Cassandra etc.)

This PR provides a stab at implementing such a reader.

### CODE REVIEW

**Comment by patriknw** [comment]
How is this related to the persistence testkit https://github.com/akka/akka/pull/25393?
It also has an inmem storage and I'm not sure we want to maintain two?

**Comment by cchepelov** [comment]
Hi @patriknw , thanks for the pointer (wasn't aware of it)

I agree that any added code is a liability and must be justified.

Skimming over #25393: 
  - @manonthegithub's efforts are much broader and ambitious in scope, and that PR seems very promising (thanks!)
  - https://github.com/akka/akka/pull/25393/files#diff-89b1f2d2476461c08f5e1f10576c98a9 InMemJournal:
    - uses concurrent data structures to store received events
    - is scoped at testing level and woven within a TestKit
    - uses the TestKitStorage structure at both the **event** and **snapshot** levels
    - does not implement the ReadJournal interface (the ones providing a Source[Event, NotUsed])
    - does not appear to affect InMemJournal itself
 - this PR:
    - is restricted to providing a readable journal, nothing else
    - uses plain messaging to store received events (into its MessageReceiver)
    - is scoped "compile", not "test" in order to remain on the same plane as the other journal reading facilities (InMemJournal is scoped "compile", why would its ReadJournal facility be scoped otherwise?)
    - affects InMemJournal itself
    - works at the Event level only (snapshots are entirely out of scope)
    - provides the ReadJournal interface (the ones providing a Source[Event, NotUsed]) including both "current" and "live" sources

TL;DR: I don't think these two PR compete and are the same thing. #25393 enables testing in detail the logistics of storing events and snapshots (and then some more). This one enables substituting a "real" heavyweight database engine for testing the logistics of jumping from one aggregate to another, or from one aggregate to its view projections.

Nevertheless, there is the question of the apparent duplication of the event storage facility. I'm not sure what would be best here…

**Comment by patriknw** [comment]
Closing this due to inactivity, and https://github.com/akka/akka/pull/26825 seems promising. Support for queries on top of that would be interesting.

**Comment by jroper** [comment]
Just voicing my support for this PR, in memory implementations of things make it very quick to get something working, allowing you to focus for example on validating a project concept, before needing to think about deploying or even selecting a database.
